lists/arraylist/arraylist.go

```diff
@@ -155,16 +152,15 @@ func (list *List[T]) Insert(index int, values ...T) {
 
 	if !list.withinRange(index) {
 		// Append
-		if index == list.size {
+		if index == len(list.elements) {
```

Comment:
Probably also worth replacing with `slices.Insert`

Reply:
In this case, we may need to create a benchmark to test the performance of `Insert()`

Reply:
To avoid performance issues after switching to `slices.Insert()`, I also create a benchmark tests. Current tests show that the performance impact is negligible.

```diff
func (list *List[T]) Insert(index int, values ...T) {
	if !list.withinRange(index) {
		// Append
		if index == len(list.elements) {
			list.Add(values...)
		}
		return
	}
+	l := len(values)
-	l := len(list.elements)
	list.growBy(len(values))
+	list.elements = slices.Insert(list.elements[:l], index, values...)
-	copy(list.elements[index+l:], list.elements[index:len(list.elements)-l])
-	copy(list.elements[index:], values)
}
```

```go
func benchmarkInsert(b *testing.B, size int) {
	for i := 0; i < b.N; i++ {
		list := New[int]()
		for n := 0; n < size; n++ {
			list.Insert(list.Size()/2, n)
		}
	}
}

func BenchmarkArrayListInsert100(b *testing.B) {
	benchmarkInsert(b, 100)
}

func BenchmarkArrayListInsert1000(b *testing.B) {
	benchmarkInsert(b, 1000)
}

func BenchmarkArrayListInsert10000(b *testing.B) {
	benchmarkInsert(b, 10000)
}

func BenchmarkArrayListInsert100000(b *testing.B) {
	benchmarkInsert(b, 100000)
}
```

```
goos: linux
goarch: amd64
pkg: github.com/emirpasic/gods/v2/lists/arraylist
cpu: Intel(R) Core(TM) i7-4790 CPU @ 3.60GHz
                        │ copy.insert.txt │         slices.insert.txt            │
                        │     sec/op      │    sec/op     vs base                │
ArrayListInsert100-8         7.561µ ± 17%   7.955µ ± 19%   +5.20% (p=0.009 n=50)
ArrayListInsert1000-8        103.8µ ± 12%   114.9µ ± 11%  +10.67% (p=0.024 n=50)
ArrayListInsert10000-8       3.792m ±  2%   3.824m ±  2%        ~ (p=0.095 n=50)
ArrayListInsert100000-8      590.9m ±  0%   592.7m ±  0%        ~ (p=0.400 n=50)
geomean                      1.152m         1.200m         +4.17%

                        │ copy.insert.txt │         slices.insert.txt             │
                        │      B/op       │     B/op      vs base                 │
ArrayListInsert100-8         1.906Ki ± 0%   1.906Ki ± 0%       ~ (p=1.000 n=50) ¹
ArrayListInsert1000-8        15.91Ki ± 0%   15.91Ki ± 0%       ~ (p=1.000 n=50) ¹
ArrayListInsert10000-8       255.9Ki ± 0%   255.9Ki ± 0%       ~ (p=0.303 n=50)
ArrayListInsert100000-8      2.000Mi ± 0%   2.000Mi ± 0%       ~ (p=0.303 n=50)
geomean                      63.14Ki        63.14Ki       +0.00%
¹ all samples are equal

                        │ copy.insert.txt │        slices.insert.txt            │
                        │    allocs/op    │ allocs/op   vs base                 │
ArrayListInsert100-8           6.000 ± 0%   6.000 ± 0%       ~ (p=1.000 n=50) ¹
ArrayListInsert1000-8          9.000 ± 0%   9.000 ± 0%       ~ (p=1.000 n=50) ¹
ArrayListInsert10000-8         13.00 ± 0%   13.00 ± 0%       ~ (p=1.000 n=50) ¹
ArrayListInsert100000-8        16.00 ± 0%   16.00 ± 0%       ~ (p=1.000 n=50) ¹
geomean                        10.29        10.29       +0.00%
¹ all samples are equal
```